Fix(CubeSQL) Add ability to filter dates inclusive of date being passed in w…

darapuk opened the pull request:
…hen using <= or >=

**Check List**
- [x] Tests has been run in packages where changes made if available
- [x] Linter has been run for changed code
- [ ] Tests for the changes have been added if not covered yet
- [ ] Docs have been added / updated if required

**Issue Reference this PR resolves**

[[#3369]](https://github.com/cube-js/cube/issues/3369) 
[[#6312]](https://github.com/cube-js/cube/issues/6312)

**Description of Changes Made (if issue reference is not provided)**

Example query:
```
EXPLAIN SELECT encounter_start from encounters WHERE encounter_start >= '2018'
```

What cube actually runs:
```
SELECT encounter_start from encounters WHERE encounter_start > '2018'
```

The problem with this is what the graphql endpoint receives, note the afterDate:
```
CubeScan: request={
  "measures": [],
  "dimensions": [
    "encounters.encounter_start"
  ],
  "segments": [],
  "filters": [
    {
      "member": "encounters.encounter_start",
      "operator": "afterDate",
      "values": [
        "2018"
      ]
    }
  ]
}
```

The problem with this is that it's not doing a GreaterThanOrEqual, it's doing a GreaterThan . Which means all SQL queries which rely on comparing a time using >= or <= are incorrect when using the SQL API endpoint.

This also introduces the `AfterOrOnDate` to the GraphQL endpoint so that users have a more accurate way of representing a time window.

darapuk commented on rust/cubesql/cubesql/src/compile/mod.rs:
```diff
@@ -4048,94 +4072,169 @@ ORDER BY \"COUNT(count)\" DESC"
+                    V1LoadRequestQueryFilterItem {
+                        member: Some("KibanaSampleDataEcommerce.order_date".to_string()),
+                        operator: Some("beforeDate".to_string()),
+                        values: Some(vec!["2021-09-06T23:59:59.999Z".to_string()]),
```
There's some code that is forcing this interval to change and I cannot seem to trace down to see where this is occuring.

paveltiunov replied:
It's a regression. Please see further comments.

paveltiunov commented on rust/cubesql/cubesql/src/compile/rewrite/rules/filters.rs:
```diff
@@ -2253,6 +2253,59 @@ impl RewriteRules for FilterRules {
                     "FilterOpOp:and",
                 ),
             ),
+            transforming_rewrite(
```
We shouldn't add new rules here. If two rules like these are introduced previous two should be adjusted to use minus one millisecond logic so all existing tests are still passing.

darapuk replied:
Ahh, I see. To make sure I'm following, my takeaway is that the regressions are coming from the `date_range` filter not matching.  This change only matches when both sides of the SQL is <= and >=. So there would need to be additional filters for <= and > and vice versa. 

One pointer I would really appreciate is how to incorporate the minus_one logic with the filter?

I.e. I want to add the following filter:
```
transforming_rewrite(
                "filter-replacer-in-date-range-inclusive-two",
                filter_op(
                    filter_op_filters(
                        filter_member(
                            "?member",
                            "FilterMemberOp:afterOrOnDate",
                            "?date_range_start",
                        ),
                        filter_member(
                            "?member",
                            "FilterMemberOp:beforeDate",
                            "?date_range_end", <-- Here, I want to call a minus_one transformation
                        ),
                    ),
                    "FilterOpOp:and",
                ),
                filter_member("?member", "FilterMemberOp:inDateRange", "?date_range"),
                self.merge_date_range("?date_range_start", "?date_range_end", "?date_range"),
            ),
```

I'm struggling to figure out what the minus_one transformation should look like here

darapuk replied:
The dumb solution I can think of is to add in a transformation param into `self.merge_date_range("?date_range_start", "?date_range_end", "?date_range")` and do the transformation in merge_date_range.

i.e.
`self.merge_date_range("?date_range_start", "?date_range_end", "?date_range", 'TransformMemberOp:StartMinusOne)`
`self.merge_date_range("?date_range_start", "?date_range_end", "?date_range", 'TransformMemberOp:StartPlusOne)`
`self.merge_date_range("?date_range_start", "?date_range_end", "?date_range", 'TransformMemberOp:EndMinusOne)`
`self.merge_date_range("?date_range_start", "?date_range_end", "?date_range", 'TransformMemberOp:EndPlusOne)`

paveltiunov replied:
@darapuk I'd actually take a slightly bit different approach. In fact `filter-replacer-in-date-range-inclusive` rule should become smart enough to match either `afterDate` or `afterOrOnDate` and `beforeDate` or `beforeOrOnDate` and coerce it to `inDateRange` with minus/plus ms logic. To avoid cartesian products of all possible inversions it should also understand if `afterDate` comes first or last. So it'll be the one rule to rule them all.

darapuk replied:
I logically understand where you're going, but need that hint for how to extract the "operation" from the SQL. In `filter-replacer-in-date-range-inclusive` the filter_member seems to match exactly on the filter type:
```
let op = match_data_node!(node_by_id, params[1], FilterMemberOp);
```

What I'm not following is how to extract which type of operator is being used then pass it into the transformer.

I.e. 
What I want to be able to write is something like:
```
            transforming_rewrite(
                "filter-replacer-in-date-range-inclusive",
                filter_op(
                    filter_op_filters(
                        filter_member(
                            "?member",
                            "?left_operator",
                            "?date_range_left",
                        ),
                        filter_member(
                            "?member",
                            "?right_operator",
                            "?date_range_right",
                        ),
                    ),
                    "FilterOpOp:and",
                ),
                filter_member("?member", "FilterMemberOp:inDateRange", "?date_range"),
                self.merge_date_range("?date_range_left", "?date_range_right", "?date_range", "?right_operator", "left_operator"),
```

is that at all close, or am I totally misunderstanding how these filters are working?

paveltiunov reviewed (changes requested):
Thanks for contributing to it! This is a big one and, and indeed very tricky! Most of the provided changes to tests are regressions. So first of all we should avoid updating existing tests. To fix all of those minus one ms logic should be extracted to matching rule for afterDate/beforeDate during time inDateRange conversion: so if you see afterDate/beforeDate operator timestamp should be adjusted by 1 millisecond during the rewrite transformation phase.

paveltiunov commented:
Here's where minus one ms logic is implemented for your reference: https://github.com/cube-js/cube/pull/7041/files#diff-bb70f8319a553cbdd6e606a1e7dec4310814ca2199e35e839258a99db1757f4aL2653

igorlukanin commented:
@darapuk Would you also be open to updating the documentation for APIs you've changing? Thanks in advance! https://cube.dev/docs/product/apis-integrations

darapuk commented:
Thanks for the review @paveltiunov! I reverted all the tests with regressions and just updated the ones which are incorrect. I realize the code is horribly inefficient and not very DRY. But at least now the code is in a "working" state.

I'm not familiar enough with how the filters work to take it the last mile unfortunately. Some examples or direction would help me push it over the line, or if anyone else wants to contribute to the PR, open to that as well!

paveltiunov commented:
@darapuk Yep. You are almost there! We just need to get rid of this duplicated code. You can parameterize `FilterMemberOp` operators using variables like `?date_range_start_op` and `?date_range_end_op`.

darapuk commented:
Thanks for that tip @paveltiunov! I believe the code is now in a bit cleaner state for a re-review.

paveltiunov reviewed (changes requested):
We're at the last mile! We just need to fix matching issues

darapuk commented on rust/cubesql/cubesql/src/compile/rewrite/rules/filters.rs:
```diff
@@ -3810,27 +3790,92 @@ impl FilterRules {
+                                resolve_time_delta(&date_range_start[0], date_range_start_op);
+                            let resolved_end_date =
+                                resolve_time_delta(&date_range_end[0], date_range_end_op);
+                            if resolved_start_date <= resolved_end_date {
```
OOF. Good catch! Updated the code to handle the left and right + the overlapping operator case.

darapuk commented on rust/cubesql/cubesql/src/compile/mod.rs:
```diff
@@ -17873,26 +17873,56 @@ ORDER BY \"COUNT(count)\" DESC"
 
     #[tokio::test]
-    async fn test_thoughtspot_filter_date_trunc_column_with_literal() {
+    async fn test_thoughtspot_filter_date_trunc_column_with_literal2() {
```
Nope, reverted

darapuk commented:
Hi @paveltiunov, 

when you get a sec, would love to get another review on this and push it through! Thanks!

paveltiunov commented:
@darapuk Looks good! Thanks for contributing this one!

siamak-haschemi commented:
Thank you also from my team @darapuk. We now can remove the millisecond fix during our SQL generation.
